parquet/src/main/java/org/apache/iceberg/parquet/ParquetFormatModel.java

```diff
@@ -0,0 +1,394 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.schema.MessageType;
+
+public class ParquetFormatModel<D, S, F> implements FormatModel<D, S> {
+  public static final String WRITER_VERSION_KEY = "parquet.writer.version";
+
+  private final Class<? extends D> registerType;
+  private final Class<S> schemaType;
+  private final ReaderFunction<D> readerFunction;
+  private final BatchReaderFunction<D, F> batchReaderFunction;
+  private final WriterFunction<S> writerFunction;
+
+  private ParquetFormatModel(
+      Class<? extends D> registerType,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      BatchReaderFunction<D, F> batchReaderFunction,
+      WriterFunction<S> writerFunction) {
+    this.registerType = registerType;
+    this.schemaType = schemaType;
+    this.readerFunction = readerFunction;
+    this.batchReaderFunction = batchReaderFunction;
+    this.writerFunction = writerFunction;
+  }
+
+  public ParquetFormatModel(Class<D> type) {
+    this(type, null, null, null, null);
+  }
+
+  public ParquetFormatModel(
+      Class<D> type,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      WriterFunction<S> writerFunction) {
+    this(type, schemaType, readerFunction, null, writerFunction);
+  }
+
+  public ParquetFormatModel(
+      Class<? extends D> returnType,
+      Class<S> schemaType,
+      BatchReaderFunction<D, F> batchReaderFunction) {
+    this(returnType, schemaType, null, (BatchReaderFunction<D, F>) batchReaderFunction, null);
+  }
+
+  @Override
+  public FileFormat format() {
+    return FileFormat.PARQUET;
+  }
+
+  @Override
+  public Class<D> type() {
+    return (Class<D>) registerType;
+  }
+
+  @Override
+  public Class<S> schemaType() {
+    return schemaType;
+  }
+
+  @Override
+  public WriteBuilder<D, S> writeBuilder(EncryptedOutputFile outputFile) {
+    return new WriteBuilderWrapper<>(outputFile, writerFunction);
+  }
+
+  @Override
+  public ReadBuilder<D, S> readBuilder(InputFile inputFile) {
+    if (readerFunction != null) {
+      return new NonBatchReaderWrapper<>(inputFile, readerFunction);
+    } else if (batchReaderFunction != null) {
+      return new BatchReaderWrapper<>(inputFile, batchReaderFunction);
+    } else {
+      throw new IllegalStateException("Either readerFunction or batchReaderFunction must be set");
+    }
+  }
+
+  @FunctionalInterface
+  public interface ReaderFunction<D> {
+    ParquetValueReader<D> read(
+        Schema schema, MessageType messageType, Map<Integer, ?> constantValues);
+  }
+
+  @FunctionalInterface
+  public interface BatchReaderFunction<D, F> {
+    VectorizedReader<D> read(
+        Schema schema,
+        MessageType messageType,
+        Map<Integer, ?> constantValues,
+        F deleteFilter,
+        Map<String, String> config);
```

Comment:
This should be removed (see previous comments).

Reply:
Let's talk about it at the other place: https://github.com/apache/iceberg/pull/12298#discussion_r2548086267